src/field_theory/minimal_polynomial.lean

```diff
@@ -0,0 +1,230 @@
+  by_contra hnz,
+  have := degree_le_of_ne_zero hx hnz _,
+  { contrapose! this,
+    exact degree_mod_by_monic_lt _ (monic hx) (ne_zero hx) },
+  { rw ← mod_by_monic_add_div p (monic hx) at hp,
+    simpa using hp }
+end
+
+/--The degree of a minimal polynomial is nonzero.-/
+lemma degree_ne_zero : degree (minimal_polynomial hx) ≠ 0 :=
+begin
+  assume deg_eq_zero,
+  have ndeg_eq_zero : nat_degree (minimal_polynomial hx) = 0,
+  { simpa using congr_arg nat_degree (eq_C_of_degree_eq_zero deg_eq_zero) },
+  have eq_one : minimal_polynomial hx = 1,
+  { rw eq_C_of_degree_eq_zero deg_eq_zero, congr,
+    simpa [ndeg_eq_zero.symm] using (monic hx).leading_coeff },
+  simpa [eq_one, aeval_def] using aeval hx
+end
+
+/--A minimal polynomial is not a unit.-/
+lemma not_is_unit : ¬ is_unit (minimal_polynomial hx) :=
+begin
+  intro H, apply degree_ne_zero hx,
+  exact degree_eq_zero_of_is_unit H
+end
+
+/--The degree of a minimal polynomial is positive.-/
+lemma degree_pos : 0 < degree (minimal_polynomial hx) :=
+degree_pos_of_ne_zero_of_nonunit (ne_zero hx) (not_is_unit hx)
+
+/--A minimal polynomial is prime.-/
+lemma prime : prime (minimal_polynomial hx) :=
+begin
+  refine ⟨ne_zero hx, not_is_unit hx, _⟩,
+  rintros p q ⟨d, h⟩,
+  have :    polynomial.aeval α β x (p*q) = 0 := by simp [h, aeval hx],
+  replace : polynomial.aeval α β x p = 0 ∨ polynomial.aeval α β x q = 0 := by simpa,
+  cases this; [left, right]; apply dvd; assumption
+end
+
+/--A minimal polynomial is irreducible.-/
+lemma irreducible : irreducible (minimal_polynomial hx) :=
+irreducible_of_prime (prime hx)
+
+/--If L/K is a field extension, and x is an element of L in the image of K,
+then the minimal polynomial of x is X - C x.-/
+@[simp] protected lemma algebra_map (a : α) (ha : is_integral α (algebra_map β a)) :
+  minimal_polynomial ha = X - C a :=
+begin
+  refine (unique ha (monic_X_sub_C a) (by simp [aeval_def]) _).symm,
+  intros q hq H,
+  rw degree_X_sub_C,
+  suffices : 0 < degree q,
+  { -- This part is annoying and shouldn't be there.
+    have q_ne_zero : q ≠ 0,
+    { apply polynomial.ne_zero_of_degree_gt this },
+    rw degree_eq_nat_degree q_ne_zero at this ⊢,
+    rw [← with_bot.coe_zero, with_bot.coe_lt_coe] at this,
+    rwa [← with_bot.coe_one, with_bot.coe_le_coe], },
+  apply degree_pos_of_root (ne_zero_of_monic hq),
+  show is_root q a,
+  apply is_field_hom.injective (algebra_map β : α → β),
+  rw [is_ring_hom.map_zero (algebra_map β : α → β), ← H],
+  convert polynomial.hom_eval₂ _ _ _ _,
+  { exact is_semiring_hom.id },
+  { apply_instance }
+end
+
+variable (β)
+/--If L/K is a field extension, and x is an element of L in the image of K,
+then the minimal polynomial of x is X - C x.-/
+lemma algebra_map' (a : α) :
+  minimal_polynomial (@is_integral_algebra_map α β _ _ _ _ _ a) =
+  X - C a :=
+minimal_polynomial.algebra_map _ _
+variable {β}
+
+/--The minimal polynomial of 0 is X.-/
+@[simp] lemma zero {h₀ : is_integral α (0:β)} :
```

Comment:
Can't one supply a proof that 0 is integral over alpha here, rather than asking for one? Similarly for 1 below.

Reply:
The downside to that is that I think `simp` won't kick in because not all proofs are equal enough for `simp`

Reply:
Yep, and for almost every use case you'll be simplifying left-to-right, and hence not need to provide the witness, as it will be inferred.